### PR TITLE
[CP] Add migrator to upgrade gradle version when conflict with Android Studio bundled Java version is detected (and 3 others)

### PR DESCRIPTION
**Pull request by gmackall**
This is a cherry pick of the following four PRs (the first one depends on changes made in the latter three).
1. https://github.com/flutter/flutter/pull/125836
2. https://github.com/flutter/flutter/pull/124771
3. https://github.com/flutter/flutter/pull/123034
4. https://github.com/flutter/flutter/pull/123644

For the first PR, 
Original issue: https://github.com/flutter/flutter/issues/122376
Cherry-pick issue: https://github.com/flutter/flutter/issues/125693

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by reidbaker** [comment]
This does not appear to be a clean cherry pick. I would not expect to see the java version stuff packages/flutter_tools/gradle/flutter.gradle in this pr.

**Comment by reidbaker** [comment]
This is bigger than I hoped it would be on the last day we can make cherry pick changes.

**Review by chunhtai — APPROVED** [review]
LGTM

**Review by andrewkolos — APPROVED** [review]
LGTM

**Comment by reidbaker** [comment]
I grabbed @Hixie to talk about this CP. 

TLDR: We should accept this CP but do so as a dot release after google IO

Factors for not accepting include: 
* This is the last day we could accept CP for the 3.10 initial release. 
* The original pr had to be reverted because of a breaking bug. 
* The [re-land pr](https://github.com/flutter/flutter/pull/126081) landed yesterday without much time to get testing on master. 
* The CP depends on multiple CLs that are not related to this problem increasing the scope/complexity. 
* Updating to flutter alone will not break users who would benefit by this pr they also need to update android studio. 

Factors for accepting: 
* The re-land pr has a broad try catch block covering the new code with non breaking fallback behavior. 
* This code technically modifies flutter build and doing that in a patch is not ideal
*  We believe there are lots of flutter users with simple apps that would benefit from the migration and are not comfortable in the gradle/agp/java area of code that will be exposed to that layer without the migration. 

Conclusion, let the code live on master for a longer period of time to get some more testing then accept for a patch to 3.10.

**Comment by eliasyishak** [comment]
Just a friendly ping from PR review incase we wanted to start this process now

**Comment by XilaiZhang** [comment]
On post submit, we are seeing a failure in [[Mac_ios flutter_gallery_ios__start_up](https://ci.chromium.org/p/flutter/builders/prod/Mac_ios%20flutter_gallery_ios__start_up)], I am rerunning the task but it does not look like a flake based on build history. Might have to revert it from stable if it does not pass on rerun. cc @CaseyHillers

**Comment by gmackall** [comment]
It looks like it succeeded on retry. 

This particular test was last marked as flakey around a month ago https://github.com/flutter/flutter/issues/125106, with what looks to me like a similar failure (but I am not very knowledgable about this test).
Recent one: https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_ios%20flutter_gallery_ios__start_up/10841/overview
Failure from 6 days ago: https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_ios%20flutter_gallery_ios__start_up/10769/overview

cc'ing @vashworth, from what I can see you've looked into the flakiness of this test somewhat recently, so if you have any input on if this is a known flake that would be great! But also understand if you're not the right person to ask

**Comment by vashworth** [comment]
Both https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_ios%20flutter_gallery_ios__start_up/10769/overview and https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_ios%20flutter_gallery_ios__start_up/10841/overview appear to be dart vm timeout flakes, which is a long standing issue across multiple iOS tests being tracked here: https://github.com/flutter/flutter/issues/121231.

There is no fix as of yet other than rerunning

**Comment by XilaiZhang** [comment]
thanks for inspecting! yeah it passed on rerun. right I only looked at the past 50 build history and didn't know it was a flake. landed on stable.

**Comment by gmackall** [comment]
> Both https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_ios%20flutter_gallery_ios__start_up/10769/overview and https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_ios%20flutter_gallery_ios__start_up/10841/overview appear to be dart vm timeout flakes, which is a long standing issue across multiple iOS tests being tracked here: #121231.
> 
> There is no fix as of yet other than rerunning

Great, thanks for the clarification!
